#Εργασία 2: **Ταξινόμηση για OCR**


> *Για την παρακάτω εργασια εχω δουλεψει σε Google Colab.*

In [1]:
import numpy as np
from IPython.display import Markdown
import matplotlib.pyplot as plt


##**2. Άνοιγμα αρχείου και προεργασία**

###Φόρτωση του Letter Recognition Data Set

*   20000 Αντικείμενα
*   26 κλάσεις: Α to Z
*   16 μεταβλητές:


In [2]:
#It will prompt you to select a file. Select and upload the file "letter-recognition.data".
#You should be able to see the name of the file once Colab has uploaded it.
from google.colab import files
uploaded = files.upload()

#χρησιμοποιούμε τη βιβλιοθήκη csv για να χωρίσουμε τα στοιχεία του αρχείου 
#και να δημιουργήσουμε έτσι μια λίστα, κάθε στοιχείο της οποίας είναι επίσης λίστα, και αντιστοιχεί σε μια γραμμή/αντικείμενο
import csv
results = []
with open('letter-recognition.data') as myfile:
  data = csv.reader(myfile, delimiter=',')
  for dataline in data:
    results.append(dataline)



Saving letter-recognition.data to letter-recognition.data


### Δημιουργία Χ και Υ πινάκων 

In [3]:
# Μετατροπή σε Numy Array        
XY=np.array(results)
X = XY[:,1:]
y = XY[:,0]
X = X.astype(float)
y_str = y

#Για να χρησιμοποιήσουμε τα y στο scikit-learn θα χρειαστεί να τα μετατρέψουμε και αυτά σε αριθμητικά. 
from sklearn.preprocessing import LabelEncoder
le = LabelEncoder().fit(y_str)
y = le.transform(y_str)

display(Markdown(
    "### Χαρακτηριστικά Letter Recognition DataSet:\n" +
    "**Αντικείμενα**: " + str(np.shape(X)[0]) + "<br>" +
    "**Μεταβλητές**: " + str(np.shape(X)[1]) + "<br>" + 
    "**Κλάσεις**: " + str(np.shape(np.unique(y))[0]) + "<br>" +
    "**Παραδείγματα δεδομένων**: "
))
import pandas as pd
df = pd.DataFrame(data= np.c_[X , y], columns=['x-box	horizontal position of box', 'y-box	vertical position of box', 'width	width of box', 'high 	height of box', 'onpix	total # on pixels', 'x-bar	mean x of on pixels in box',
                                               'y-bar	mean y of on pixels in box', 'x2bar	mean x variance', 'y2bar	mean y variance', 'xybar	mean x y correlation', 'x2ybr	mean of x * x * y', 'xy2br	mean of x * y * y',
                                               'x-ege	mean edge count left to right', 'xegvy	correlation of x-ege with y', 'y-ege	mean edge count bottom to top',
                                               'yegvx	correlation of y-ege with x ', 'Class'])
df.head(15)






### Χαρακτηριστικά Letter Recognition DataSet:
**Αντικείμενα**: 20000<br>**Μεταβλητές**: 16<br>**Κλάσεις**: 26<br>**Παραδείγματα δεδομένων**: 

x-box\thorizontal position of box  y-box\tvertical position of box  \
0                                 2.0                              8.0   
1                                 5.0                             12.0   
2                                 4.0                             11.0   
3                                 7.0                             11.0   
4                                 2.0                              1.0   
5                                 4.0                             11.0   
6                                 4.0                              2.0   
7                                 1.0                              1.0   
8                                 2.0                              2.0   
9                                11.0                             15.0   
10                                3.0                              9.0   
11                                6.0                             13.0   
12                                4.0                              9.0   
13                                6.0                              9.0   
14                                5.0                              9.0   

    width\twidth of box  high \theight of box  onpix\ttotal # on pixels  \
0                   3.0                   5.0                       1.0   
1                   3.0                   7.0                       2.0   
2                   6.0                   8.0                       6.0   
3                   6.0                   6.0                       3.0   
4                   3.0                   1.0                       1.0   
5                   5.0                   8.0                       3.0   
6                   5.0                   4.0                       4.0   
7                   3.0                   2.0                       1.0   
8                   4.0                   4.0                       2.0   
9                  13.0                   9.0                       7.0   
10                  5.0                   7.0                       4.0   
11                  4.0                   7.0                       4.0   
12                  6.0                   7.0                       6.0   
13                  8.0                   6.0                       9.0   
14                  5.0                   7.0                       6.0   

    x-bar\tmean x of on pixels in box  y-bar\tmean y of on pixels in box  \
0                                 8.0                               13.0   
1                                10.0                                5.0   
2                                10.0                                6.0   
3                                 5.0                                9.0   
4                                 8.0                                6.0   
5                                 8.0                                8.0   
6                                 8.0                                7.0   
7                                 8.0                                2.0   
8                                10.0                                6.0   
9                                13.0                                2.0   
10                                8.0                                7.0   
11                                6.0                                7.0   
12                                7.0                                8.0   
13                                7.0                                8.0   
14                                6.0                               11.0   

    x2bar\tmean x variance  y2bar\tmean y variance  \
0                      0.0                     6.0   
1                      5.0                     4.0   
2                      2.0                     6.0   
3                      4.0                     6.0   
4                      6.0                     6.0   
5                      6.0                     9.0   
6                      

### Διαχωρισμός training-test set

In [4]:
X_train = X[0:16000,:]
y_train = y[0:16000]

X_test = X[16000:,:]
y_test = y[16000:]


###Τυπώστε το μέσο, τη διασπορά, το μέγιστο και το ελάχιστο κάθε στήλης του x_train και του x_test.

In [5]:
display(Markdown(
    "**Το Μέσο της κάθε μεταβλητης για X_train** : " + str(np.mean(X_train,0))  + "<br>" +
    "**Το Μέσο της κάθε μεταβλητης για X_test** : " + str(np.mean(X_test,0))  + "<br>" +
    "**Η Διασπορα της κάθε μεταβλητης για X_train** : " + str(np.var(X_train,0))  + "<br>" +
    "**Η Διασπορα της κάθε μεταβλητης για X_test** : " + str(np.var(X_test,0)) + "<br>" +
    #"**Η Τυπική Απόκλιση της κάθε μεταβλητης για X_train** : " + str(np.std(X_train,0))  + "<br>" +
    #"**Η Τυπική Απόκλιση της κάθε μεταβλητης για X_test** : " + str(np.std(X_test,0)) + "<br>" +
    "**Το Min της κάθε μεταβλητης για X_train** : " + str(np.min(X_train,0)) + "<br> **και το Max** : " + str(np.max(X_train,0)) + "<br>" +
    "**Το Min της κάθε μεταβλητης για X_test** : " + str(np.min(X_test,0)) + "<br> **και το Max** : " + str(np.max(X_test,0))
    
))

**Το Μέσο της κάθε μεταβλητης για X_train** : [4.0201875 7.0289375 5.1173125 5.365875  3.500125  6.892625  7.5124375
 4.6273125 5.170375  8.286625  6.47125   7.927125  3.04875   8.3439375
 3.682     7.79625  ]<br>**Το Μέσο της κάθε μεταβλητης για X_test** : [4.037   7.06175 5.14    5.39875 3.52875 6.9175  7.4525  4.63375 5.21175
 8.26375 6.385   7.9365  3.0355  8.3185  3.73075 7.821  ]<br>**Η Διασπορα της κάθε μεταβλητης για X_train** : [ 3.64215496 10.91522512  4.01280028  5.11913548  4.80574998  4.11872061
  5.44659531  7.35216653  5.68722236  6.17834611  6.98004844  4.29281423
  5.48687344  2.4025195   6.611626    2.57073594]<br>**Η Διασπορα της κάθε μεταβλητης για X_test** : [ 3.732131   10.93593694  4.2399      5.09074844  4.76567344  4.04769375
  5.24574375  7.03861094  5.58991194  6.24718594  6.684775    4.47246775
  5.25473975  2.35055775  6.49925444  2.796959  ]<br>**Το Min της κάθε μεταβλητης για X_train** : [0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 1.]<br> **και το Max** : [15. 15. 15. 15. 15. 15. 15. 15. 15. 15. 15. 15. 15. 15. 15. 15.]<br>**Το Min της κάθε μεταβλητης για X_test** : [0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 2. 0. 1. 0. 0.]<br> **και το Max** : [14. 15. 14. 14. 15. 15. 15. 15. 15. 15. 15. 15. 15. 14. 15. 15.]

##**3.2. Γραμμική Ταξινόμηση**


###**Λογιστική παλινδρόμηση**

In [6]:
from sklearn.pipeline import Pipeline
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import precision_score, recall_score, f1_score
from sklearn.preprocessing import MinMaxScaler


log_reg_pipeline=Pipeline([('Scaler',  MinMaxScaler()),
            ('LogRegClassifier', LogisticRegression(max_iter=1000))])

params = { 
           'LogRegClassifier__C': np.logspace(3, -2, 6) 
         }

grid = GridSearchCV(estimator=log_reg_pipeline, cv = 5, param_grid=params, verbose=4)
grid.fit(X_train,y_train)

y_test_pred = grid.predict(X_test)

display(Markdown(
    "### Αποτελέσματα Λογιστικής Παλινδρόμησης:\n" +
    "**Best lambda**: {:.3f}".format(grid.best_params_['LogRegClassifier__C'])  + "<br>" +
    "**Precision**: {:.3f}".format(precision_score(y_test, y_test_pred, average='weighted'))  + "<br>" +
    "**Recall**: {:.3f}".format(recall_score(y_test, y_test_pred, average='weighted'))  + "<br>" +
    "**Test Set F1**: {:.3f}".format(f1_score(y_test, y_test_pred, average='weighted')) 
    ))


Fitting 5 folds for each of 6 candidates, totalling 30 fits
[CV 1/5] END ........LogRegClassifier__C=1000.0;, score=0.773 total time=  22.9s
[CV 2/5] END ........LogRegClassifier__C=1000.0;, score=0.768 total time=  23.4s
[CV 3/5] END ........LogRegClassifier__C=1000.0;, score=0.777 total time=  22.5s
[CV 4/5] END ........LogRegClassifier__C=1000.0;, score=0.762 total time=  20.7s
[CV 5/5] END ........LogRegClassifier__C=1000.0;, score=0.782 total time=  25.2s
[CV 1/5] END .........LogRegClassifier__C=100.0;, score=0.772 total time=  19.1s
[CV 2/5] END .........LogRegClassifier__C=100.0;, score=0.770 total time=  20.0s
[CV 3/5] END .........LogRegClassifier__C=100.0;, score=0.776 total time=  19.8s
[CV 4/5] END .........LogRegClassifier__C=100.0;, score=0.765 total time=  21.4s
[CV 5/5] END .........LogRegClassifier__C=100.0;, score=0.782 total time=  20.9s
[CV 1/5] END ..........LogRegClassifier__C=10.0;, score=0.770 total time=  14.2s
[CV 2/5] END ..........LogRegClassifier__C=10.0;,

### Αποτελέσματα Λογιστικής Παλινδρόμησης:
**Best lambda**: 100.000<br>**Precision**: 0.772<br>**Recall**: 0.774<br>**Test Set F1**: 0.771

###**SVM Γραμμικού Πυρήνα**

In [7]:
#Import svm model
from sklearn.svm import SVC

LinearSVM_pipeline=Pipeline([('Scaler',  MinMaxScaler()),
            ('SVMClassifier', SVC(kernel='linear'))])

params = {
            'SVMClassifier__C': np.logspace(3, 0, 4)
         }

grid = GridSearchCV(estimator=LinearSVM_pipeline, cv = 5, param_grid=params, verbose=4)
grid.fit(X_train,y_train)


y_test_pred = grid.predict(X_test)


display(Markdown(
    "### Αποτελέσματα Γραμμικου SVM:\n" +
    "**Best C**: {:.3f}".format(grid.best_params_['SVMClassifier__C'])  + "<br>" +
    "**Precision**: {:.3f}".format(precision_score(y_test, y_test_pred, average='weighted'))  + "<br>" +
    "**Recall**: {:.3f}".format(recall_score(y_test, y_test_pred, average='weighted'))  + "<br>" +
    "**Test Set F1**: {:.3f}".format(f1_score(y_test, y_test_pred,average='weighted')) 
    ))

Fitting 5 folds for each of 4 candidates, totalling 20 fits
[CV 1/5] END ...........SVMClassifier__C=1000.0;, score=0.859 total time=   8.6s
[CV 2/5] END ...........SVMClassifier__C=1000.0;, score=0.847 total time=   9.7s
[CV 3/5] END ...........SVMClassifier__C=1000.0;, score=0.851 total time=   8.3s
[CV 4/5] END ...........SVMClassifier__C=1000.0;, score=0.841 total time=   8.2s
[CV 5/5] END ...........SVMClassifier__C=1000.0;, score=0.858 total time=   8.4s
[CV 1/5] END ............SVMClassifier__C=100.0;, score=0.856 total time=   3.7s
[CV 2/5] END ............SVMClassifier__C=100.0;, score=0.848 total time=   3.7s
[CV 3/5] END ............SVMClassifier__C=100.0;, score=0.849 total time=   3.7s
[CV 4/5] END ............SVMClassifier__C=100.0;, score=0.839 total time=   3.7s
[CV 5/5] END ............SVMClassifier__C=100.0;, score=0.860 total time=   3.7s
[CV 1/5] END .............SVMClassifier__C=10.0;, score=0.853 total time=   3.2s
[CV 2/5] END .............SVMClassifier__C=10.0;,

### Αποτελέσματα Γραμμικου SVM:
**Best C**: 1000.000<br>**Precision**: 0.838<br>**Recall**: 0.835<br>**Test Set F1**: 0.835

##**3.2. Μη-γραμμική ταξινόμηση**


###**k-Nearest Neighbors**

In [8]:
from sklearn.neighbors import KNeighborsClassifier

KNN_pipeline=Pipeline([('Scaler',  MinMaxScaler()),
            ('kNNClassifier', KNeighborsClassifier())])

params = { 
           'kNNClassifier__n_neighbors':   range(1,7),
           'kNNClassifier__weights':   ['distance']
         }

grid = GridSearchCV(estimator=KNN_pipeline, cv = 5, param_grid=params, verbose=4)

grid.fit(X_train, y_train) 
y_test_pred = grid.predict(X_test)

display(Markdown(
    "### Αποτελέσματα k-nearest neighbors classification:\n" +
    "**Best k**: {:.3f}".format(grid.best_params_['kNNClassifier__n_neighbors'])  + "<br>" +
    "**Precision**: {:.3f}".format(precision_score(y_test, y_test_pred, average='weighted'))  + "<br>" +
    "**Recall**: {:.3f}".format(recall_score(y_test, y_test_pred, average='weighted'))  + "<br>" +
    "**Test Set F1**: {:.3f}".format(f1_score(y_test, y_test_pred,average='weighted'))
))

Fitting 5 folds for each of 6 candidates, totalling 30 fits
[CV 1/5] END kNNClassifier__n_neighbors=1, kNNClassifier__weights=distance;, score=0.950 total time=   0.6s
[CV 2/5] END kNNClassifier__n_neighbors=1, kNNClassifier__weights=distance;, score=0.948 total time=   0.5s
[CV 3/5] END kNNClassifier__n_neighbors=1, kNNClassifier__weights=distance;, score=0.953 total time=   0.5s
[CV 4/5] END kNNClassifier__n_neighbors=1, kNNClassifier__weights=distance;, score=0.948 total time=   0.4s
[CV 5/5] END kNNClassifier__n_neighbors=1, kNNClassifier__weights=distance;, score=0.953 total time=   0.4s
[CV 1/5] END kNNClassifier__n_neighbors=2, kNNClassifier__weights=distance;, score=0.950 total time=   0.6s
[CV 2/5] END kNNClassifier__n_neighbors=2, kNNClassifier__weights=distance;, score=0.947 total time=   0.5s
[CV 3/5] END kNNClassifier__n_neighbors=2, kNNClassifier__weights=distance;, score=0.952 total time=   0.5s
[CV 4/5] END kNNClassifier__n_neighbors=2, kNNClassifier__weights=distance;,

### Αποτελέσματα k-nearest neighbors classification:
**Best k**: 4.000<br>**Precision**: 0.958<br>**Recall**: 0.958<br>**Test Set F1**: 0.958

###**SVM Πυρήνα Ακτινικής Βάσης**

In [9]:
RBF_SVM_pipeline=Pipeline([('Scaler',  MinMaxScaler()),
            ('RBFSVMClassifier', SVC(kernel='rbf'))])

params = {
            'RBFSVMClassifier__C': np.logspace(3, 0, 4),
            'RBFSVMClassifier__gamma': np.logspace(1, -1, 6)
         }

grid = GridSearchCV(estimator=RBF_SVM_pipeline, cv = 5, param_grid=params, verbose=4)
grid.fit(X_train,y_train)

y_test_pred = grid.predict(X_test)


display(Markdown(
    "### Αποτελέσματα RBF SVM:\n" +
    "**Best C**: {:.3f}".format(grid.best_params_['RBFSVMClassifier__C'])  + "<br>" +
    "**Best gamma**: {:.3f}".format(grid.best_params_['RBFSVMClassifier__gamma'])  + "<br>" +
    "**Precision**: {:.3f}".format(precision_score(y_test, y_test_pred, average='weighted'))  + "<br>" +
    "**Recall**: {:.3f}".format(recall_score(y_test, y_test_pred, average='weighted'))  + "<br>" +
    "**Test Set F1**: {:.3f}".format(f1_score(y_test, y_test_pred,average='weighted')) 
    ))

Fitting 5 folds for each of 24 candidates, totalling 120 fits
[CV 1/5] END RBFSVMClassifier__C=1000.0, RBFSVMClassifier__gamma=10.0;, score=0.976 total time=   8.3s
[CV 2/5] END RBFSVMClassifier__C=1000.0, RBFSVMClassifier__gamma=10.0;, score=0.974 total time=   8.3s
[CV 3/5] END RBFSVMClassifier__C=1000.0, RBFSVMClassifier__gamma=10.0;, score=0.978 total time=   8.2s
[CV 4/5] END RBFSVMClassifier__C=1000.0, RBFSVMClassifier__gamma=10.0;, score=0.968 total time=   8.2s
[CV 5/5] END RBFSVMClassifier__C=1000.0, RBFSVMClassifier__gamma=10.0;, score=0.973 total time=   8.1s
[CV 1/5] END RBFSVMClassifier__C=1000.0, RBFSVMClassifier__gamma=3.9810717055349722;, score=0.971 total time=   4.9s
[CV 2/5] END RBFSVMClassifier__C=1000.0, RBFSVMClassifier__gamma=3.9810717055349722;, score=0.971 total time=   4.9s
[CV 3/5] END RBFSVMClassifier__C=1000.0, RBFSVMClassifier__gamma=3.9810717055349722;, score=0.975 total time=   4.9s
[CV 4/5] END RBFSVMClassifier__C=1000.0, RBFSVMClassifier__gamma=3.98107

### Αποτελέσματα RBF SVM:
**Best C**: 10.000<br>**Best gamma**: 10.000<br>**Precision**: 0.979<br>**Recall**: 0.979<br>**Test Set F1**: 0.979

###**Random Forests**

In [10]:
from sklearn.ensemble import RandomForestClassifier

clf = RandomForestClassifier(n_estimators=500, random_state=0)
clf.fit(X_train,y_train)

y_test_pred = clf.predict(X_test)

display(Markdown(
    "### Αποτελέσματα Random Forests:\n" +
    "**Precision**: {:.3f}".format(precision_score(y_test, y_test_pred, average='weighted'))  + "<br>" +
    "**Recall**: {:.3f}".format(recall_score(y_test, y_test_pred, average='weighted'))  + "<br>" +
    "**Test Set F1**: {:.3f}".format(f1_score(y_test, y_test_pred,average='weighted')) 
    ))


### Αποτελέσματα Random Forests:
**Precision**: 0.966<br>**Recall**: 0.965<br>**Test Set F1**: 0.965

###**Extremely Randomized Trees**

In [11]:
from sklearn.ensemble import ExtraTreesClassifier

clf = ExtraTreesClassifier(n_estimators=500, random_state=0)
clf.fit(X_train,y_train)

y_test_pred = clf.predict(X_test)

display(Markdown(
    "### Αποτελέσματα Extremely Randomized Trees:\n" +
    "**Precision**: {:.3f}".format(precision_score(y_test, y_test_pred, average='weighted'))  + "<br>" +
    "**Recall**: {:.3f}".format(recall_score(y_test, y_test_pred, average='weighted'))  + "<br>" +
    "**Test Set F1**: {:.3f}".format(f1_score(y_test, y_test_pred,average='weighted')) 
    ))

### Αποτελέσματα Extremely Randomized Trees:
**Precision**: 0.973<br>**Recall**: 0.973<br>**Test Set F1**: 0.973

##**4. Αξιολόγηση αποτελεσμάτων**
> **Παρουσίαση συμπερασμάτων:**

* Αρχίκα, όσον αφορά την **γραμμική ταξινόμηση** που υλοποιήσαμε χρησιμοποιώντας **Λογιστική Παλινδρόμηση** και **SVM Γραμμικού Πυρήνα** παρατηρώντας τα *Precision* και *Recall* (ή το *F1 Score* το οποίο συνδυάζει τα 2 παραπάνω κριτήρια) συμπεραίνουμε ότι κανένα από τα δύο μοντέλα δεν πραγματοποίησε αρκετα καλή ταξινόμηση καθώς οι τιμές τους, 0.77 και 0.83 έκαστος, είναι μέτριες είδικα αν συνυπολογίσουμε το γεγονός οτι ο χρονος που χρειαστηκαν για να τελειωσουν ηταν *6 λεπτα* (Λογιστική Παλινδρόμηση) και *1 λεπτο και 55 δευτερόλεπτα* (SVM Γραμμικού Πυρήνα).
*   Συνεχίζοντας, όσον αφορα την **μη-γραμμική ταξινόμηση** που υλοποιήσαμε χρησιμοποιώντας **k-Nearest Neighbors**, **SVM Πυρήνα Ακτινικής Βάσης**, **Random Forests** και **Extremely Randomized Trees** παρατηρόυμε ότι όλα τα παραπάνω μοντέλα έχουν υψηλές τιμές *precision* και *recall* (και *F1 Score*) με τιμές που κυμαινονται από το 0.95 εως και 0.98, πράγμα που σημαίνει ότι οι ταξινομητές μας ταξινόμούν αρκετά καλα όλα τα αντικείμενα στην κατάλληλη κλάση (κατάλληλο χαρακτήρα).
*   Παρα την υψηλή επίδοση που έχει ο αλγόριθμος **SVM Πυρήνα Ακτινικής Βάσης**, παρατηρείται **μεγάλη καθυστέρηση** καθώς χρειάζεται *περίπου 12 λεπτά* για να ολοκλήρωθεί. 
*   Αξίζει να σημείωθει το γεγονός οτι οι αλγόριθμοι: **k-Nearest Neighbors**, **Random Forests** και **Extremely Randomized Trees** πετυχαίνουν τόσο **υψηλή επίδοση** όσο και **χαμηλές καθυστερήσεις**, ολοκληρώνοντας μόνο μέσα σε *22 δευτερόλεπτα* (k-Nearest Neighbors), *12 δευτερόλεπτα* (Random Forests) και *10 δευτερόλεπτα* (Extremely Randomized Trees).


*Για αυτο ακριβώς το λόγο παρακάτω έχω επιλέξει ως καλύτερο αλγόριθμο τον **Extremely Randomized Trees** καθώς έχει και υψηλή επιδοση και χαμηλές καθυστερήσεις.*




###**Classification Report**

In [12]:
from sklearn.metrics import classification_report

print("Classification report for classifier: "+ str(clf) +"\n" )
print(classification_report(y_test, y_test_pred))


Classification report for classifier: ExtraTreesClassifier(n_estimators=500, random_state=0)

              precision    recall  f1-score   support

           0       0.99      0.99      0.99       156
           1       0.92      0.98      0.95       136
           2       0.99      0.98      0.99       142
           3       0.93      0.99      0.96       167
           4       0.97      0.97      0.97       152
           5       0.96      0.97      0.97       153
           6       0.97      0.98      0.97       164
           7       0.98      0.91      0.95       151
           8       0.96      0.96      0.96       165
           9       0.98      0.95      0.97       148
          10       0.97      0.97      0.97       146
          11       0.99      0.97      0.98       157
          12       0.97      1.00      0.99       144
          13       0.99      0.94      0.96       166
          14       0.95      0.99      0.97       139
          15       0.98      0.95      0.

Η κλάση που ταξινομείται χειρότερα είναι η 7 (που αντιστοιχεί στο χαρακτήρα Η) καθώς το precision (0.94), recall (0.91), f1-score (0.93) είναι τα χαμηλότερα σε σχέση με των υπολοίπων χαρακτήρων.